### PR TITLE
Fix/ Postpone application activation if registration is not complete yet

### PR DESCRIPTION
**Pull request by LitvinenkoIra**
Fixes https://github.com/smartdevicelink/sdl_core/issues/3549

This PR is **[ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Testing Plan
Script from https://github.com/smartdevicelink/sdl_atf_test_scripts/pull/2470

### Summary
There was an issue when the SDL receives SDLActivateAppRequest from HMI and activates the app before the registration is completed and NONE HMI level is assigned. In this case, after the registration is complete SDL will assign NONE hmi level for the app that was in FULL so additional steps will be required to activate the application one more time.

To prevent such a situation additional check for app HMI level has been added during the app activation (SDLActivateAppRequest):
if the app HMI level is INVALID_ENUM - it means that the registration isn't complete yet - so we need to postpone the activation.
After the registration is completed and HMI level NONE is assigned - need to check if we have the postponed activation for the app and if yes - start the postponed activation.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by LitvinenkoIra** [comment]
@theresalech This PR is ready for Livio review. Thank you!

File: src/components/application_manager/src/state_controller_impl.cc [hunk 1]
```diff
@@ -955,6 +957,15 @@ void StateControllerImpl::OnStateChanged(ApplicationSharedPtr app,
 
   if (new_state->hmi_level() == mobile_apis::HMILevel::HMI_NONE) {
     app->ResetDataInNone();
+    if (mobile_apis::HMILevel::INVALID_ENUM == old_state->hmi_level()) {
+      const uint32_t app_id = app->app_id();
+      const uint32_t corr_id =
+          postponed_activation_controller_.GetPendingActivationCorrId(app_id);
+      if (corr_id > 0) {
+        app_mngr_.GetPolicyHandler().OnActivateApp(app_id, corr_id);
```
Comment (iCollin) [inline src/components/application_manager/src/state_controller_impl.cc:965]: Why is onActivateApp called before these two lines are run?
```
app_mngr_.OnHMIStateChanged(app->app_id(), old_state, new_state);
app->usage_report().RecordHmiStateChanged(new_state->hmi_level());
```

File: src/components/application_manager/src/state_controller_impl.cc [hunk 2]
```diff
@@ -955,6 +957,15 @@ void StateControllerImpl::OnStateChanged(ApplicationSharedPtr app,
 
   if (new_state->hmi_level() == mobile_apis::HMILevel::HMI_NONE) {
     app->ResetDataInNone();
+    if (mobile_apis::HMILevel::INVALID_ENUM == old_state->hmi_level()) {
```
Comment (iCollin) [inline src/components/application_manager/src/state_controller_impl.cc:960]: Can you please explain when this code will run when `GetDefaultHmiLevel` returns a status other than `NONE`?

Comment (AKalinich-Luxoft) [reply src/components/application_manager/src/state_controller_impl.cc:960]: @iCollin I believe this code will not be executed if app transits from INVALID to whatever except NONE. It is a good point and I think this block should be moved to the end of function

Comment (AKalinich-Luxoft) [reply src/components/application_manager/src/state_controller_impl.cc:960]: @iCollin I believe this code will not be executed if app transits from INVALID to whatever except NONE. It is a good point and I think this block should be moved to the end of function

File: src/components/application_manager/src/state_controller_impl.cc (hunk 1 above)
Comment (AKalinich-Luxoft) [reply src/components/application_manager/src/state_controller_impl.cc:965]: @iCollin as I can see `OnActivateApp` indirectly triggers the same function `OnStateChanged` too.
Probably, it make sense to move these lines before

File: src/components/application_manager/src/state_controller_impl.cc (hunk 2 above)
Comment (JackLivio) [reply src/components/application_manager/src/state_controller_impl.cc:960]: @AKalinich-Luxoft At the end of this function i suggest this is the implemented fix:

```
  if (mobile_apis::HMILevel::INVALID_ENUM == old_state->hmi_level() && app_mngr_.GetDefaultHmiLevel(app) == new_state->hmi_level()) {
    const uint32_t app_id = app->app_id();
    const uint32_t corr_id =
        postponed_activation_controller_.GetPendingActivationCorrId(app_id);
    if (corr_id > 0) {
      app_mngr_.GetPolicyHandler().OnActivateApp(app_id, corr_id);
      postponed_activation_controller_.RemoveAppToActivate(app_id);
    }
  }
```

Comment (JackLivio) [reply src/components/application_manager/src/state_controller_impl.cc:960]: This should address the possibility for a default hmi level other than NONE

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/sdl_activate_app_request.cc [hunk 3]
```diff
@@ -186,6 +207,18 @@ void SDLActivateAppRequest::Run() {
...
+          "yet, postpone activation");
+      auto& postponed_activation_ctrl = application_manager_.state_controller()
+                                            .GetPostponedActivationController();
+      postponed_activation_ctrl.AddAppToActivate(application_id,
+                                                 correlation_id());
+      return;
+    }
     policy_handler_.OnActivateApp(application_id, correlation_id());
     return;
   } else if (app_to_activate->is_cloud_app()) {
```
Comment (iCollin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/sdl_activate_app_request.cc:224]: @AKalinich-Luxoft there are some cases where this delay logic will not be applied, such as if the app is a cloud app. I think the other call to `policy_handler_.OnActivateApp` in `SDLActivateAppRequest::on_event` should have the postponed_activation_ctrl logic as well.

Comment (AKalinich-Luxoft) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/sdl_activate_app_request.cc:224]: @iCollin fixed in https://github.com/smartdevicelink/sdl_core/pull/3557/commits/1b59bec2a67c251ed6455496501a91a64c14b4b5

File: src/components/application_manager/src/state_controller_impl.cc (hunk 1 above)
Comment (AKalinich-Luxoft) [reply src/components/application_manager/src/state_controller_impl.cc:972]: @iCollin fixed in https://github.com/smartdevicelink/sdl_core/pull/3557/commits/4605e9859af606ef434484e8c9ea094918e4d365

File: src/components/application_manager/src/state_controller_impl.cc (hunk 2 above)
Comment (AKalinich-Luxoft) [reply src/components/application_manager/src/state_controller_impl.cc]: @iCollin @JackLivio fixed in https://github.com/smartdevicelink/sdl_core/pull/3557/commits/4605e9859af606ef434484e8c9ea094918e4d365

**Pull request merged**